### ISSUE TITLE
DescribeCertificate did not contain DomainValidationOptions after 10 tries

### ISSUE THREAD

**Issue by afgallo** [aws/copilot-cli#2602]
Hi there!

I tried to deploy my front-end service today and got the following error:

> Received response status [FAILED] from custom resource. Message returned: DescribeCertificate did not contain DomainValidationOptions after 10 tries. (Log: /aws/lambda/xxx-CertificateValidationFunction-b9AmhJj5xZyy/2021/07/13/[$LATEST]de2507a6f95044e5a03a4f5d864416c0) (RequestId: 73db11eb-4028-4d9b-8645-c1f85e714122)

You can find below my `manifest.yml`:

```name: front-end
type: Load Balanced Web Service

http:
  path: '/'
  healthcheck:
    path: '/hc'
    success_codes: '200'
    healthy_threshold: 3
    unhealthy_threshold: 2
    interval: 15s
    timeout: 3s
  stickiness: false

image:
  build: Dockerfile
  port: 4200

cpu: 256 
memory: 512 
count: 1
exec: true

network:
  vpc:
    placement: 'public'

environments:
  dev:
    count: 1
    http:
      alias: dev.myapp.domain.com
  staging:
    count: 1
    cpu: 1024
    http:
      alias: staging.myapp.domain.com
  prod:
    cpu: 512
    memory: 1024
    range: 1-3
    # requests: 10000
    # response_time: 2s
    cpu_percentage: 70
    memory_percentage: 80
    http:
      alias: myapp.domain.com
```

I've checked a few things that may help you guys try and repro this bug:

1. SSL certs are created and issued successfully although they don't get deleted after the stack gets rolled back
2. Cloudformation seems to get stuck creating `EnvControllerAction`
3. The lambda functions used to validate the cert seems to be created successfully too

Any help would be much appreciated as this issue is impacting our ability to go live with our application using AWS Copilot.

Thanks!

**Comment by efekarakus** [comment]
Hi @afgallo !

Which version of Copilot are you using?

Would you mind sharing with us how the terminal looks like when the error occurs with `copilot deploy`? so that we can validate  that it's indeed the EnvControllerAction that's failing or another action

**Comment by afgallo** [comment]
Hi @efekarakus, I'm sorry I forgot to add those details. I'm using version 1.8.3 and below is a screenshot of the error:

**During execution**
<img width="991" alt="bug01" src="https://user-images.githubusercontent.com/9999415/125378796-ba1c8280-e3d2-11eb-8273-fb5cdfbb0650.png">

**After execution**
<img width="1118" alt="bug02" src="https://user-images.githubusercontent.com/9999415/125378756-a244fe80-e3d2-11eb-86e6-69fab5305e78.png">

**Comment by efekarakus** [comment]
Gotcha thank you! It does look like it's the certificate validation lambda that's failing.  
Which region is this happening in? I wonder if there is a transient issue with ACM.

Would you mind also using the AWS CLI: https://docs.aws.amazon.com/cli/latest/reference/acm/describe-certificate.html
`aws acm describe-certificate --certificate-arn <ssl cert arn>` during the deployment to see the output (for DomainValidationOptions) and pasting it here for us 🙏 

Apologies for the inconvenience 🙇

**Comment by afgallo** [comment]
This is happening in `ap-southeast-2`, @efekarakus.

<img width="1524" alt="bug03" src="https://user-images.githubusercontent.com/9999415/125381529-69f3ef00-e3d7-11eb-8b51-d027c7cf227b.png">

Here is the output of the above command during deployment:

```
{
    "Certificate": {
        "CertificateArn": "arn:aws:acm:ap-southeast-2:xxxx:certificate/xxxx",
        "DomainName": "xx.xx.xx.app",
        "SubjectAlternativeNames": [
            "xx.xx.xx.app",
            "*.xx.xx.xx.app"
        ],
        "DomainValidationOptions": [
            {
                "DomainName": "xx.xx.xx.app",
                "ValidationDomain": "xx.xx.xx.app",
                "ValidationStatus": "SUCCESS",
                "ResourceRecord": {
                    "Name": "_a5e08c90151e01ab9a26dbb0c6548339.xx.xx.xx.app.",
                    "Type": "CNAME",
                    "Value": "_3112c62f427c5296e28638c020c563cd.lblqlwmygg.acm-validations.aws."
                },
                "ValidationMethod": "DNS"
            },
            {
                "DomainName": "*.xx.xx.xx.app",
                "ValidationDomain": "*.xx.xx.xx.app",
                "ValidationStatus": "SUCCESS",
                "ResourceRecord": {
                    "Name": "_a5e08c90151e01ab9a26dbb0c6548339.xx.xx.xx.app.",
                    "Type": "CNAME",
                    "Value": "_3112c62f427c5296e28638c020c563cd.lblqlwmygg.acm-validations.aws."
                },
                "ValidationMethod": "DNS"
            }
        ],
        "Serial": "0e:24:2d:0e:fc:12:bb:54:9c:de:cc:7f:bd:c9:81:1f",
        "Subject": "CN=xx.xx.xx.app",
        "Issuer": "Amazon",
        "CreatedAt": "2021-07-13T12:35:07+10:00",
        "IssuedAt": "2021-07-13T12:35:25+10:00",
        "Status": "ISSUED",
        "NotBefore": "2021-07-13T10:00:00+10:00",
        "NotAfter": "2022-08-12T09:59:59+10:00",
        "KeyAlgorithm": "RSA-2048",
        "SignatureAlgorithm": "SHA256WITHRSA",
        "InUseBy": [],
        "Type": "AMAZON_ISSUED",
        "KeyUsages": [
            {
                "Name": "DIGITAL_SIGNATURE"
            },
            {
                "Name": "KEY_ENCIPHERMENT"
            }
        ],
        "ExtendedKeyUsages": [
            {
                "Name": "TLS_WEB_SERVER_AUTHENTICATION",
                "OID": "1.3.6.1.5.5.7.3.1"
            },
            {
                "Name": "TLS_WEB_CLIENT_AUTHENTICATION",
                "OID": "1.3.6.1.5.5.7.3.2"
            }
        ],
   ...
```

Interestingly, after the stack rolled back, this is what I see in the Certificate Manager UI (two identical certs):

<img width="1525" alt="bug04" src="https://user-images.githubusercontent.com/9999415/125381895-11712180-e3d8-11eb-830e-648ad197f0e9.png">

**Comment by efekarakus** [comment]
OK! I think we found out the root cause!

**Can you remove the `alias` field from the `dev` and `staging` environments?**  
The reason why it's failing is that by default we will set an alias value of `dev.myapp.domain.com` and `staging.myapp.domain.com` and we don't expect those values to be set in the manifest file again by a user. 

So if you update the manifest to be like this:
<details>
  <summary>Manifest file</summary>

```yaml
type: Load Balanced Web Service

http:
  path: '/'
  healthcheck:
    path: '/hc'
    success_codes: '200'
    healthy_threshold: 3
    unhealthy_threshold: 2
    interval: 15s
    timeout: 3s
  stickiness: false

image:
  build: Dockerfile
  port: 4200

cpu: 256 
memory: 512 
count: 1
exec: true

network:
  vpc:
    placement: 'public'

environments:
  dev:
    count: 1
  staging:
    count: 1
    cpu: 1024
  prod:
    cpu: 512
    memory: 1024
    range: 1-3
    # requests: 10000
    # response_time: 2s
    cpu_percentage: 70
    memory_percentage: 80
    http:
      alias: myapp.domain.com
```
</details>

It should work! The service will be available for dev at `dev.myapp.domain.com`, staging at `staging.myapp.domain.com` and prod at `myapp.domain.com`.

On our end we should update our code so that we just ignore customer value if it's set to the default generated cert domain.
https://github.com/aws/copilot-cli/blob/e070b5fd0844a56df59ec8ce69a95e3a3011eb3d/cf-custom-resources/lib/dns-cert-validator.js#L112-L115
https://github.com/aws/copilot-cli/blob/e070b5fd0844a56df59ec8ce69a95e3a3011eb3d/cf-custom-resources/lib/dns-cert-validator.js#L142-L156
Otherwise the `expectedValidationOptionsNum` is equal to 3 instead of 2.

**Comment by efekarakus** [comment]
(Reopening until @afgallo validates that the issue is fixed)

**Comment by afgallo** [comment]
Hi @efekarakus and @iamhopaul123

I can confirm the issue is now fixed. Although I agree with @iamhopaul123 this wasn't expected behaviour, I think the confusion was I thought the default aliases would carry the service name with them in the format of `${SvcName}.${EnvName}.${AppName}.${DomainName}` which I'm sure I've seen in the docs somewhere. My mistake but I'm glad you guys added that extra check on your side. Thanks a lot again!

**Comment by iamhopaul123** [comment]
Hi @afgallo. Glad the workaround worked for this issue! Totally understandable for setting the default aliases. It is an edge case that we should definitely consider (either support it or error out early). Thank you so much for reporting this issue!

**Comment by afgallo** [comment]
Hi @efekarakus and @iamhopaul123 I think my thoughts were right .... even though the deployment succeed after having removed the http alias, I can only reach my front-end service at `front-end.dev.myapp.domain.com` and not at `dev.myapp.domain.com`. Any thoughts?

**Comment by iamhopaul123** [comment]
hello @afgallo. Did you add `alias: dev.myapp.domain.com` when you deployed your `front-end`? And the current behavior is if the alias is specified then the old default one won't work anymore...If I understand correctly you want both `front-end.dev.myapp.domain.com` and `dev.myapp.domain.com` work for the same service?

**Comment by afgallo** [comment]
Yes I did originally but @efekarakus suggested I removed the alias so I am confused now

> OK! I think we found out the root cause!
> 
> **Can you remove the `alias` field from the `dev` and `staging` environments?**
> The reason why it's failing is that by default we will set an alias value of `dev.myapp.domain.com` and `staging.myapp.domain.com` and we don't expect those values to be set in the manifest file again by a user.
> 

I actually don't need `front-end.dev.myapp.domain.com` at all. I just need the `dev.myapp.domain.com`

**Comment by afgallo** [comment]
> hello @afgallo. Did you add `alias: dev.myapp.domain.com` when you deployed your `front-end`? And the current behavior is if the alias is specified then the old default one won't work anymore...If I understand correctly you want both `front-end.dev.myapp.domain.com` and `dev.myapp.domain.com` work for the same service?

I was jus thinking through this @iamhopaul123. Would it work if I do the first deployment with the default alias and then do a second deploy with the alias I need which is `dev.myapp.domain.com`?

**Comment by efekarakus** [comment]
Hi @afgallo 👋 

Sorry for the churn! You're right the recommendation does end up resulting to only `front-end.dev.myapp.domain.com` 😞.   
The PR #2607 does fix this issue, we could set the alias `dev.myapp.domain.com` and it works. 

Few possible alternatives while waiting on the new fix:
1. Use different `alias` values for `dev` and `staging` environments:
   ```yaml
   environments:
     dev:
        http:
          alias: 'beta.myapp.domain.com'
     stage:
       http:
          alias: 'gamma.myapp.domain.com' # or 'preprod.myapp.domain.com'
   ```
   This is not ideal but when we release v1.9.0 you can change it back to `dev.myapp.domain.com`.
   
2. You can build a local version of `copilot` by running `make` from the commit that has the fix: https://github.com/aws/copilot-cli/tree/d0d42cada0bdd3596b279a7c2be4993152c4f67f

**Comment by afgallo** [comment]
Hi @efekarakus when is 1.9.0 scheduled for release?

Thanks!

**Comment by efekarakus** [comment]
@afgallo We are tentatively looking at 7/29!

**Comment by Lou1415926** [comment]
The fix is now released in [v1.9.0](https://github.com/aws/copilot-cli/releases/tag/v1.9.0)! 🎊
